common/parallel/TreeCrawler.go

```diff
@@ -101,29 +154,117 @@ func (c *crawler) workerLoop(ctx context.Context, wg *sync.WaitGroup, workerInde
 	var err error
 	mayHaveMore := true
 	for mayHaveMore && ctx.Err() == nil {
-		mayHaveMore, err = c.processOneDirectory(ctx, workerIndex)
+		mayHaveMore, err = c.processOneDirectoryWithAutoPacer(ctx, workerIndex)
 		if err != nil {
 			c.output <- CrawlResult{err: err}
 			// output the error, but we don't necessarily stop the enumeration (e.g. it might be one unreadable dir)
 		}
 	}
 }
 
-func (c *crawler) processOneDirectory(ctx context.Context, workerIndex int) (bool, error) {
+// receiveOnTqueue receive dirs enumerated by source and append to unstartedDirs.
+func (c *crawler) receiveOnTqueue() {
+	const maxQueueDirectories = 1000 * 1000
+	// Wait for any new entries on channel.
+	for tDir := range c.tqueue {
+
+		// Lets put backpressure to source to slow down, otherwise c.unstartedDirs will keep growing and
+		// and cause memory pressure on system.
+		for len(c.unstartedDirs) > maxQueueDirectories {
+			time.Sleep(10 * time.Millisecond)
+		}
+		c.cond.L.Lock()
+		if tDir != nil {
```

Comment:
also consider changing the for loop like this, the reason being
1. Avoid sleeping on every iteration (as explained in the comment)
2. We cannot safely do len(c.unstartedDirs) outside the lock. "-race" will complain I believe.

        for tDir := range c.tqueue {
                c.cond.L.Lock()
                //
                // Lets put backpressure on source to slow down, otherwise c.unstartedDirs will keep growing
                // and and cause memory pressure on system.
                // Once unstartedDirs crosses the limit we let it drain to 80% of the limit to avoid pointelessely
                // sleeping on every other iteration.
                // 20% of maxQueueDirectories is 200K directories. Even the fastest target will take multiple
                // seconds to process these many directories, so we can safely add a largish sleep.
                //
                if len(c.unstartedDirs) > maxQueueDirectories {
                        for len(c.unstartedDirs) > maxQueueDirectories*0.8 {
                                c.cond.L.Unlock()
                                time.Sleep(1000 * time.Millisecond)
                                c.cond.L.Lock()
                        }
                }
                if tDir != nil {
                        c.unstartedDirs = append(c.unstartedDirs, tDir)
                }
                c.cond.L.Unlock()
                c.cond.Broadcast()
        }